docs/architecture/adrs/002-streaming-response-aggregation.md

```diff
@@ -0,0 +1,43 @@
+    1. Generated texts in output frames with detections are still expected to be contiguous, meaning the end user can still piece all of the frames together. This aligns with the way various text generation servers provide output.
+2. In order to achieve (1) for detectors that may work with different chunkers, e.g. paragraph vs. sentence, results will not be streamed back/returned until the “slowest” detector has processed. “Slowest” may not necessarily just be in terms of time but in terms of amount of text processed (i.e. largest `processed_index`).  
+
+### Response aggregation strategy
+
+1. Only chunker calls will provide indices of the text generation server frames processed (`start_index` and `processed_index`). Detector calls are expected to be unary, and responses only indicate the detections on any given chunk from a chunker. Detectors are expected to still give a reply (i.e. empty array) if there are no detections on a given chunk.
+    1. On any detector response, the orchestrator will have to associate the chunker indices with each detection response. The orchestrator will update any span start/end on detection responses with chunker indices to be able to indicate to the end user where the detection was in the overall stream.
+2. The orchestrator will have to maintain detection responses that are not ready to be streamed back to the user, such as with queues.
+3. Initial primary aggregation strategy via maximum `processed_index`
+    1. Before aggregation and the streaming back of any frame to the end user, all detector queues will be checked for at least one response. This ensures that no results will be streamed back to the user before detectors (and corresponding chunkers) have responded.
+    2. The maximum `processed_index` for frames at the beginning of each queue will be used for reference. For the other detector queues, responses will be aggregated based on `processed_index` up to the reference maximum `processed_index`.
+    3. To ensure there are no premature responses e.g. maximum `processed_index` is 42 and another queue has responses at `processed_index` 5, 17, and 34, responses will be “awaited” for that detector until a response with `processed_index` reaches 42 or more. If more than 42, the last response will be left on the queue.
+    4. Once all the responses for up to a certain maximum `processed_index` are identified, the detections will be aggregated into a list to be returned in the `token_classification_results` field of the frame to be streamed back. The text of that frame will of course correspond to the text of the frame that had the maximum `processed_index`.
+    5. Once a frame is returned/streamed back, the process will repeat to look for at least one queued response from each detector for comparison, until each detector has processed to the end of the stream.
+4. The aggregation strategy can be configurable, even if only one is available and implemented to begin with.
```

Comment:
We should write it as an implementation / design of the streaming logic inside orchestrator. Since currently it looks like these would be configurable by the user via API.